### PR TITLE
np.trim_zeros

### PR DESCRIPTION
**Pull request by sungraek**
Add support for `np.trim_zeros`. It currently does not handle nested list or tuple input well.
#4074

### CODE REVIEW

**Comment by gmarkall** [comment]
Thanks very much for the PR, I've queued it for review.

File: numba/np/arraymath.py [hunk 1]
```diff
@@ -4251,3 +4251,31 @@ def impl(a, b):
...
+def np_trim_zeros(a, trim='fb'):
+    if not isinstance(a, (types.Array, types.Sequence, types.Tuple)):
+        raise TypeError('The first argument must be an array-like')
+    if hasattr(a, 'ndim') and a.ndim > 1:
+        raise TypingError('array must be 1D')
+
+    def impl(a, trim='fb'):
+        a_ = np.asarray(a)
+        first = 0
+        if 'f' in trim:
```
Comment (EPronovost) [inline numba/np/arraymath.py:4266]: From [this line](https://github.com/numpy/numpy/blob/v1.18.1/numpy/lib/function_base.py#L1608) it seems the numpy implementation works with upper and lower case letters for `trim`

```
>>> np.trim_zeros(np.array([0, 1, 2, 3, 0, 0]), trim="FB")
array([1, 2, 3])
```

File: numba/tests/test_np_functions.py [hunk 2]
```diff
@@ -3621,6 +3625,49 @@ def test_cross2d_exceptions(self):
...
 
+    def test_trim_zeros(self):
+
+        def arrays():
+            yield np.array([])
+            yield np.zeros(5)
+            yield np.zeros(1)
+            yield np.array([1, 2, 3])
+            yield np.array([0, 1, 2, 3])
+            yield np.array([0, 1, 2, 0, 0])
```
Comment (EPronovost) [inline numba/tests/test_np_functions.py:3636]: Could we test arrays that have non-zero values that are not ints (e.g. `np.array([0., 1., 2., np.nan, 0.])`)?
We could also test other data types that work with the numpy version (e.g. `np.array(["Hello", "world"])`).

File: numba/tests/test_np_functions.py [hunk 3]
```diff
@@ -3621,6 +3625,49 @@ def test_cross2d_exceptions(self):
...
+
+        pyfunc = np_trim_zeros
+        cfunc = jit(nopython=True)(pyfunc)
```
Comment (EPronovost) [inline numba/tests/test_np_functions.py:3639]: Can we also test when `trim != "fb"`?

**Comment by EPronovost** [comment]
Nice work! You'll probably want to update [the numpy supported features documentation](https://github.com/numba/numba/blob/master/docs/source/reference/numpysupported.rst) as part of your PR as well.

**Comment by sungraek** [comment]
I added support for both lower and upper case arguments and some new test cases as you mentioned.

File: numba/np/arraymath.py [hunk 4]
```diff
@@ -4251,3 +4251,32 @@ def impl(a, b):
...
 
     return impl
+
+
+@overload(np.trim_zeros)
+def np_trim_zeros(a, trim='fb'):
+    if not isinstance(a, (types.Array, types.Sequence, types.Tuple)):
+        raise TypeError('The first argument must be an array-like')
+    if hasattr(a, 'ndim') and a.ndim > 1:
+        raise TypingError('array must be 1D')
```
Comment (EPronovost) [inline numba/np/arraymath.py:4261]: Should we check that `trim` is a string or string literal?

Comment (sungraek) [reply numba/np/arraymath.py:4261]: It seems like `np.trim_zeros(arr)` `np.trim_zeros(arr, 'f')` and `np.trim_zeros(arr, trim)` all produces expected output. 
Would we still need to add a check?

Comment (EPronovost) [reply numba/np/arraymath.py:4261]: It would improve the error message if someone does something like `np.trim_zeros(arr, 1)`.

Comment (sungraek) [reply numba/np/arraymath.py:4261]: Oh, I see. I will add the check

**Comment by ionikha** [comment]
Hi, i'm working on a project that uses trim_zeros and numba, do you know why this wasn't added?
It looks good to go?

**Comment by gmarkall** [comment]
@ionikha Thanks for the ping, I think it just fell by the wayside and never got a review from a maintainer (even though @EPronovost very kindly gave it a review). I've added this to the agenda for the dev meeting tomorrow to see if we can kickstart it.

**Comment by gmarkall** [comment]
Assignment from triage meeting: @gmarkall to merge master and address any review comments, @guilhermeleobas to review subsequently.

**Comment by guilhermeleobas** [comment]
@gmarkall, if needed, we can switch and I can merge `main` and address any comments. Just let me know

**Comment by esc** [comment]
@gmarkall @guilhermeleobas -- can I ask for an update here? What is the status of this?

**Comment by gmarkall** [comment]
I'll put this on my near-term todo list so that it can get done.

**Comment by guilhermeleobas** [comment]
@gmarkall, I've created https://github.com/numba/numba/pull/9074 which is built on top of this PR with git conflicts fixed.

**Comment by esc** [comment]
Is replaced by #9074

**Pull request closed without merging**